cmd/clusterctl/pkg/client/repository/repository_local_unix.go

```diff
@@ -0,0 +1,211 @@
+	configVariablesClient config.VariablesClient
+	basepath              string
+	providerName          string
+	defaultVersion        string
+	componentsPath        string
+}
+
+var _ Repository = &localRepository{}
+
+// DefaultVersion returns the default version for the local repository.
+func (r *localRepository) DefaultVersion() string {
+	return r.defaultVersion
+}
+
+// RootPath returns the empty string as it is not applicable to local repositories.
+func (r *localRepository) RootPath() string {
+	return ""
+}
+
+// ComponentsPath returns the path to the components file for the local repository.
+func (r *localRepository) ComponentsPath() string {
+	return r.componentsPath
+}
+
+// GetFile returns a file for a given provider version.
+func (r *localRepository) GetFile(version, fileName string) ([]byte, error) {
+	var err error
+
+	if version == "latest" {
+		version, err = r.getLatestRelease()
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get the latest release")
+		}
+	} else if version == "" {
+		version = r.defaultVersion
+	}
+
+	absolutePath := filepath.Join(r.basepath, r.providerName, version, r.RootPath(), fileName)
+
+	f, err := os.Stat(absolutePath)
+	if err != nil {
+		return nil, errors.Errorf("failed to read file %q from local release %s", absolutePath, version)
+	}
+	if f.IsDir() {
+		return nil, errors.Errorf("invalid path: file %q is actually a directory %q", fileName, absolutePath)
+	}
+	content, err := ioutil.ReadFile(absolutePath)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to read file %q from local release %s", absolutePath, version)
+	}
+	return content, nil
+
+}
+
+// GetVersions returns the list of versions that are available for a local repository.
+func (r *localRepository) GetVersions() ([]string, error) {
+	// get all the sub-directories under {basepath}/{provider-name}/
+	releasesPath := filepath.Join(r.basepath, r.providerName)
+	files, err := ioutil.ReadDir(releasesPath)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to list release directories")
+	}
+	versions := []string{}
+	for _, f := range files {
+		if !f.IsDir() {
+			continue
+		}
+		r := f.Name()
+		sv, err := version.ParseSemantic(r)
+		if err != nil {
+			// discard releases with tags that are not a valid semantic versions (the user can point explicitly to such releases)
+			continue
+		}
+		if sv.PreRelease() != "" || sv.BuildMetadata() != "" {
+			// discard pre-releases or build releases (the user can point explicitly to such releases)
+			continue
+		}
+		versions = append(versions, r)
+	}
+	return versions, nil
+}
+
+// newLocalRepository returns a new localRepository.
+func newLocalRepository(providerConfig config.Provider, configVariablesClient config.VariablesClient) (*localRepository, error) {
+	url, err := url.Parse(providerConfig.URL())
+	if err != nil {
+		return nil, errors.Wrap(err, "invalid url")
+	}
+	absPath := url.Path
+	if url.RawPath != "" {
+		absPath = url.RawPath
+	}
```

Comment:
Looking at https://golang.org/pkg/net/url/#URL, it sounds like we should call `absPath := url.EscapedPath()` and not check Path vs RawPath. Is there something specific you need from RawPath?

Reply:
Done.